File: presentation/src/main/java/cleanarcpro/brightowusu/com/cleanarcproj/di/components/AppComponent.java
package cleanarcpro.brightowusu.com.cleanarcproj.di.components;

import javax.inject.Singleton;

import cleanarcpro.brightowusu.com.cleanarcproj.di.modules.AppModule;
import cleanarcpro.brightowusu.com.cleanarcproj.di.modules.InteractorsModule;
import cleanarcpro.brightowusu.com.cleanarcproj.di.modules.NetworkModule;
import cleanarcpro.brightowusu.com.cleanarcproj.di.modules.RepositoryModule;
import cleanarcpro.brightowusu.com.cleanarcproj.domain.abstractions.repository.IUserRepository;
import dagger.Component;
import okhttp3.OkHttpClient;

@Singleton
@Component(modules = {RepositoryModule.class, NetworkModule.class})
public interface AppComponent {

    OkHttpClient getOKOkHttpClient();

    IUserRepository getIUserRepository();

}

File: presentation/src/main/java/cleanarcpro/brightowusu/com/cleanarcproj/di/components/AppComponent.kt
package cleanarcpro.brightowusu.com.cleanarcproj.di.components;

import android.content.BroadcastReceiver
import cleanarcpro.brightowusu.com.cleanarcproj.di.modules.*
import cleanarcpro.brightowusu.com.cleanarcproj.domain.abstractions.repository.IUserRepository
import com.squareup.picasso.Picasso
import dagger.Component
import kotlinx.coroutines.ExperimentalCoroutinesApi
import kotlinx.coroutines.channels.BroadcastChannel
import retrofit2.Retrofit
import javax.inject.Singleton

@Singleton
@Component(modules = [
    RepositoryModule::class,
    NetworkModule::class,
    RetrofitModule::class,
    ImageFetcherModule::class,
    DatabaseModule::class
])
interface AppComponent {

    // Provides a handle on this below class, which can be used by other components
    fun getIUserRepository(): IUserRepository

    fun getRetrofit(): Retrofit

    fun getPicasso(): Picasso

    fun getNetworkChangeReceiver(): BroadcastReceiver

    @ExperimentalCoroutinesApi
    fun getNetworkBroadcastChannel(): BroadcastChannel<Boolean>
}
